src/frequenz/sdk/microgrid/_graph.py

```diff
@@ -492,7 +492,7 @@ def refresh_from(
             _provisional.validate()
         except Exception as err:
             raise InvalidGraphError(
-                "Cannot populate component graph from provided input!"
+                f"Failed to initialise component graph: {err}"
             ) from err
```

Comment:
When doing `from err`, Python will chain both exceptions, so you should be able to see the details of the underlying error too, so printing `err` explicitly is redundant. Are you not seeing the properly chained exceptions?

Reply:
At least not in tests. Here is an example of what gets displayed in tests:
```
FAILED tests/microgrid/test_graph.py::Test_MicrogridComponentGraph::test_refresh_from_api - frequenz.sdk.microgrid._graph.InvalidGraphError: Cannot populate component graph from provided input!
```

Reply:
But that should show you the backtrace, which should include where the sub-exception was raised too

Reply:
It does show it in the backtrace. I feel the error message by itself should be thorough and not generic.

But your call.

Reply:
Faking an exception to trigger this:
```diff
diff --git a/src/frequenz/sdk/microgrid/_graph.py b/src/frequenz/sdk/microgrid/_graph.py
index e433fa5..fddc09e 100644
--- a/src/frequenz/sdk/microgrid/_graph.py
+++ b/src/frequenz/sdk/microgrid/_graph.py
@@ -521,6 +521,7 @@ class _MicrogridComponentGraph(ComponentGraph):
 
     def validate(self) -> None:
         """Check that the component graph contains valid microgrid data."""
+        raise Exception("Fake exception")
         self._validate_graph()
         self._validate_graph_root()
         self._validate_grid_endpoint()
```

I get:
```console
$ pytest -x tests/microgrid/test_graph.py
=================================================================================================================================================== test session starts ===================================================================================================================================================
platform linux -- Python 3.11.4, pytest-7.4.0, pluggy-1.2.0
rootdir: /home/luca/devel/sdk
configfile: pyproject.toml
plugins: mock-3.11.1, anyio-3.7.1, asyncio-0.21.1, time-machine-2.12.0
asyncio: mode=Mode.AUTO
collected 38 items                                                                                                                                                                                                                                                                                                        

tests/microgrid/test_graph.py F<actor_restart_limit> Resetting restart limit to None


======================================================================================================================================================== FAILURES =========================================================================================================================================================
_________________________________________________________________________________________________________________________________________ TestComponentGraph.test_without_filters _________________________________________________________________________________________________________________________________________

self = <frequenz.sdk.microgrid._graph._MicrogridComponentGraph object at 0x7f290aca1f50>, components = {Component(component_id=1, category=<ComponentCategory.GRID: 1>, type=None), Component(component_id=3, category=<ComponentCategory.METER: 2>, type=None)}, connections = {Connection(start=1, end=3)}
correct_errors = None

    def refresh_from(
        self,
        components: Set[Component],
        connections: Set[Connection],
        correct_errors: Optional[Callable[["_MicrogridComponentGraph"], None]] = None,
    ) -> None:
        """Refresh the graph from the provided list of components and connections.
    
        This will completely overwrite the current graph data with the provided
        components and connections.
    
        Args:
            components: components to include in the graph, provided as pairs of
                the form `(component_id, component_category)`
            connections: connections to include in the graph, provided as pairs
                of component IDs describing the start and end of the connection
            correct_errors: callback that, if set, will be invoked if the
                provided graph data is in any way invalid (it will attempt to
                correct the errors by inferring what the correct data should be)
    
        Raises:
            InvalidGraphError: if the provided `components` and `connections`
                do not form a valid component graph and `correct_errors` does
                not fix it.
        """
        if not all(component.is_valid() for component in components):
            raise InvalidGraphError(f"Invalid components in input: {components}")
        if not all(connection.is_valid() for connection in connections):
            raise InvalidGraphError(f"Invalid connections in input: {connections}")
    
        new_graph = nx.DiGraph()
        for component in components:
            new_graph.add_node(component.component_id, **asdict(component))
    
        new_graph.add_edges_from(connections)
    
        # check if we can construct a valid ComponentGraph
        # from the new NetworkX graph data
        _provisional = _MicrogridComponentGraph()
        _provisional._graph = new_graph  # pylint: disable=protected-access
        if correct_errors is not None:
            try:
                _provisional.validate()
            except InvalidGraphError as err:
                _logger.warning("Attempting to fix invalid component data: %s", err)
                correct_errors(_provisional)
    
        try:
>           _provisional.validate()

src/frequenz/sdk/microgrid/_graph.py:492: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <frequenz.sdk.microgrid._graph._MicrogridComponentGraph object at 0x7f290aca14d0>

    def validate(self) -> None:
        """Check that the component graph contains valid microgrid data."""
>       raise Exception("Fake exception")
E       Exception: Fake exception

src/frequenz/sdk/microgrid/_graph.py:524: Exception

The above exception was the direct cause of the following exception:

self = <tests.microgrid.test_graph.TestComponentGraph object at 0x7f290ab1ba90>

    def test_without_filters(self) -> None:
        _graph_implementation = gr._MicrogridComponentGraph()
        graph: gr.ComponentGraph = _graph_implementation
    
        assert graph.components() == set()
        assert graph.connections() == set()
        with pytest.raises(
            KeyError,
            match="Component 1 not in graph, cannot get predecessors!",
        ):
            graph.predecessors(1)
        with pytest.raises(
            KeyError,
            match="Component 1 not in graph, cannot get successors!",
        ):
            graph.successors(1)
    
        # simplest valid microgrid: a grid endpoint and a meter
>       _graph_implementation.refresh_from(
            components={
                Component(1, ComponentCategory.GRID),
                Component(3, ComponentCategory.METER),
            },
            connections={Connection(1, 3)},
        )

tests/microgrid/test_graph.py:110: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <frequenz.sdk.microgrid._graph._MicrogridComponentGraph object at 0x7f290aca1f50>, components = {Component(component_id=1, category=<ComponentCategory.GRID: 1>, type=None), Component(component_id=3, category=<ComponentCategory.METER: 2>, type=None)}, connections = {Connection(start=1, end=3)}
correct_errors = None

    def refresh_from(
        self,
        components: Set[Component],
        connections: Set[Connection],
        correct_errors: Optional[Callable[["_MicrogridComponentGraph"], None]] = None,
    ) -> None:
        """Refresh the graph from the provided list of components and connections.
    
        This will completely overwrite the current graph data with the provided
        components and connections.
    
        Args:
            components: components to include in the graph, provided as pairs of
                the form `(component_id, component_category)`
            connections: connections to include in the graph, provided as pairs
                of component IDs describing the start and end of the connection
            correct_errors: callback that, if set, will be invoked if the
                provided graph data is in any way invalid (it will attempt to
                correct the errors by inferring what the correct data should be)
    
        Raises:
            InvalidGraphError: if the provided `components` and `connections`
                do not form a valid component graph and `correct_errors` does
                not fix it.
        """
        if not all(component.is_valid() for component in components):
            raise InvalidGraphError(f"Invalid components in input: {components}")
        if not all(connection.is_valid() for connection in connections):
            raise InvalidGraphError(f"Invalid connections in input: {connections}")
    
        new_graph = nx.DiGraph()
        for component in components:
            new_graph.add_node(component.component_id, **asdict(component))
    
        new_graph.add_edges_from(connections)
    
        # check if we can construct a valid ComponentGraph
        # from the new NetworkX graph data
        _provisional = _MicrogridComponentGraph()
        _provisional._graph = new_graph  # pylint: disable=protected-access
        if correct_errors is not None:
            try:
                _provisional.validate()
            except InvalidGraphError as err:
                _logger.warning("Attempting to fix invalid component data: %s", err)
                correct_errors(_provisional)
    
        try:
            _provisional.validate()
        except Exception as err:
>           raise InvalidGraphError(
                "Cannot populate component graph from provided input!"
            ) from err
E           frequenz.sdk.microgrid._graph.InvalidGraphError: Cannot populate component graph from provided input!

src/frequenz/sdk/microgrid/_graph.py:494: InvalidGraphError
-------------------------------------------------------------------------------------------------------------------------------------------------- Captured stdout setup --------------------------------------------------------------------------------------------------------------------------------------------------
<actor_restart_limit> Changing the restart limit from None to 0
================================================================================================================================================= short test summary info =================================================================================================================================================
FAILED tests/microgrid/test_graph.py::TestComponentGraph::test_without_filters - frequenz.sdk.microgrid._graph.InvalidGraphError: Cannot populate component graph from provided input!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
==================================================================================================================================================== 1 failed in 0.18s ====================================================================================================================================================
```

Which has the original exception (it is super verbose, yes).

I mean, we could also add it in the new exception, but if we chain exceptions many times the exception message will become illegible too, having too much context, so it is not ideal either.

Reply:
Can you show how are you calling pytest and what's the output, it is very weird if it is not showing the original error in the backtrace

Reply:
I do see it in the backtrace :)

>but if we chain exceptions many times the exception message will become illegible too, having too much context, so it is not ideal either.

Fair enough

Reply:
Lemme know if this PR should be closed.

Reply:
> I do see it in the backtrace :)

You mean in mine or yours? `pytest` takes backtraces to the moon and show a big chunk of the code in the backtrace, not just the file:line, so I understand it could be easy to miss if you are not used to it.

![image](https://github.com/frequenz-floss/frequenz-sdk-python/assets/1031485/21c05790-26cf-419b-b33d-7503d8c744d9)

If it is in yours that you don't see, I'm still curious and would like to know what's going on :-/

> Lemme know if this PR should be closed.

My vote goes to close, but @shsms approved in the past, so I would like to hear from him too, in case I'm missing something.

Reply:
In mine. I already mentioned it earlier here: https://github.com/frequenz-floss/frequenz-sdk-python/pull/602#discussion_r1305636337

Reply:
My feeling is that this would make debugging much faster, to provide some context in the message,  or best of all print a log message,  so we don't have to parse the whole exception,  looking for a needle in a haystack.

But it is subjective,  and although I have encountered these, I haven't really paid conscious attention to these errors, so I guess I was able to understand and debug them in their current form as well.  So maybe I shouldn't be saying much without gaining more conscious experience.

Reply:
> so we don't have to parse the whole exception, looking for a needle in a haystack.

This was the motivation behind the change. Having it mentioned clearly upfront helps.

> or best of all print a log message

This works too.

Reply:
> >or best of all print a log message
>
> This works too.

Done. Removed error description from exception.

Reply:
Thanks @sahas-subramanian-frequenz !

I'll wait for a final review from @llucax .